packages/core/src/config/parse.ts

```diff
@@ -1185,6 +1222,127 @@ const parseContractVariables = (
   return parsedConfigVariables
 }
 
+const parseArrayConstructorArg = (
+  input: ParamType,
+  constructorArgValue: UserConfigVariable
+): ParsedConfigVariable[] => {
+  if (!Array.isArray(constructorArgValue)) {
+    throw new InputError(
+      `Expected array for ${input.name} but got ${typeof constructorArgValue}`
+    )
+  }
+
+  if (input.arrayLength !== -1) {
+    if (constructorArgValue.length !== input.arrayLength) {
+      throw new InputError(
+        `Expected array of length ${input.arrayLength} for ${input.name} but got array of length ${constructorArgValue.length}`
+      )
+    }
+  }
+
+  const parsedValues: ParsedConfigVariable = []
+  for (const element of constructorArgValue) {
+    parsedValues.push(
+      parseAndValidateConstructorArg(input.arrayChildren, element)
+    )
+  }
+
+  return parsedValues
+}
+
+export const parseStructConstructorArg = (
+  paramType: ParamType,
+  constructorArgValue: UserConfigVariable
+) => {
+  if (typeof constructorArgValue !== 'object') {
+    throw new InputError(
+      `Expected object for ${
+        paramType.name
+      } but got ${typeof constructorArgValue}`
+    )
+  }
+
+  const parsedValues: ParsedConfigVariable = {}
+  for (const [key, value] of Object.entries(constructorArgValue)) {
+    const inputChild = paramType.components.find((component) => {
+      return component.name === key
+    })
+    if (inputChild === undefined) {
+      throw new InputError(
+        `User entered incorrect member in ${paramType.name}: ${key}`
+      )
+    }
+    parsedValues[key] = parseAndValidateConstructorArg(inputChild, value)
+  }
+
+  return parsedValues
+}
+
+const parseAndValidateConstructorArg = (
+  input: ParamType,
+  constructorArgValue: UserConfigVariable
+): ParsedConfigVariable => {
```

Comment:
I split the constructor arg validation out into it's own function which follows a similar recursive pattern to the parseAndValidateVariable function. I don't think it's worth trying to reuse the `recursiveLayoutIterator` for this though because the context is meaningfully different. So I think it would just add a lot of unnecessary bloat to it.